pkg/controller/controller_utils.go

```diff
@@ -594,6 +636,56 @@ func maxContainerRestarts(pod *api.Pod) int {
 	return maxRestarts
 }
 
+// ControlActivePods first filters out inactive pods, then it tries to take
+// control of the remaining pods. It returns the pods that are active and are
+// controlled by the controller.
+func ControlActivePods(podControl PodControlInterface, controllerObject api.ObjectMeta, controllerVersion, controllerKind string, pods []api.Pod) []*api.Pod {
+	var result []*api.Pod
+	for i := range pods {
+		pod := pods[i]
+		if !IsPodActive(pod) {
+			glog.V(4).Infof("Ignoring inactive pod %v/%v in state %v, deletion time %v",
+				pod.Namespace, pod.Name, pod.Status.Phase, pod.DeletionTimestamp)
+			continue
+		}
+		if controlled := ControlPod(podControl, controllerObject, controllerVersion, controllerKind, pod); !controlled {
+			glog.V(4).Infof("Failed to control pod %v/%v", pod.Namespace, pod.Name)
+			continue
+		}
+		result = append(result, &pod)
+	}
+	return result
+}
+
+// ControlPod sends patch to take control of the pod if doesn't have a
+// controller yet. It returns true if the pod end up being controlled by the
+// controller.
+func ControlPod(podControl PodControlInterface, controllerObject api.ObjectMeta, controllerVersion, controllerKind string, pod api.Pod) bool {
```

Comment:
I don't like combining logic that checks if pod is controlled with the one that sets the controllerRef - in a design we agreed that we should take the adoption logic as deterministic as possible, hence we want to adopt Pods in the oldest controller. It'll be easier to write if those two logics were separate.